pkg/reconciler/kubernetes/internal/patcher.go

```diff
@@ -0,0 +1,256 @@
+	backOffPeriod = 1 * time.Second
+	// how many times we can retry before back off
+	triesBeforeBackOff = 1
+)
+
+func newPatcher(info *resource.Info, helper *resource.Helper) *Patcher {
+	var openapiSchema openapi.Resources
+
+	return &Patcher{
+		Mapping:       info.Mapping,
+		Helper:        helper,
+		Overwrite:     true,
+		BackOff:       clockwork.NewRealClock(),
+		Force:         false,
+		Cascade:       true,
+		Timeout:       time.Duration(0),
+		GracePeriod:   -1,
+		OpenapiSchema: openapiSchema,
+		Retries:       0,
+	}
+}
+
+type Patcher struct {
+	Mapping *meta.RESTMapping
+	Helper  *resource.Helper
+
+	Overwrite bool
+	BackOff   clockwork.Clock
+
+	Force       bool
+	Cascade     bool
+	Timeout     time.Duration
+	GracePeriod int
+
+	// If set, forces the patch against a specific resourceVersion
+	ResourceVersion *string
+
+	// Number of retries to make if the patch fails with conflict
+	Retries int
+
+	OpenapiSchema openapi.Resources
+}
+
+// Patch tries to patch an OpenAPI resource. On success, returns the merge patch as well
+// the final patched object. On failure, returns an error.
+func (p *Patcher) Patch(current runtime.Object, modified []byte,
+	namespace, name string) ([]byte, runtime.Object, error) {
+	var getErr error
+
+	patchBytes, patchObject, err := p.patchSimple(current, modified, namespace, name)
+
+	if p.Retries == 0 {
+		p.Retries = maxPatchRetry
+	}
+
+	for i := 1; i <= p.Retries && errors.IsConflict(err); i++ {
+		if i > triesBeforeBackOff {
+			p.BackOff.Sleep(backOffPeriod)
+		}
+
+		current, getErr = p.Helper.Get(namespace, name)
+		if getErr != nil {
+			return nil, nil, getErr
+		}
+
+		patchBytes, patchObject, err = p.patchSimple(current, modified, namespace, name)
+	}
+
+	if err != nil && (errors.IsConflict(err) || errors.IsInvalid(err)) && p.Force {
+		patchBytes, patchObject, err = p.deleteAndCreate(current, modified, namespace, name)
+	}
+
+	return patchBytes, patchObject, err
+}
+
+func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, namespace, name string) ([]byte, runtime.Object, error) {
+	// Serialize the current configuration of the object from the server.
+	current, err := runtime.Encode(unstructured.UnstructuredJSONScheme, obj)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Retrieve the original configuration of the object from the annotation.
+	original, err := util.GetOriginalConfiguration(obj)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var (
+		patchType       types.PatchType
+		patch           []byte
+		lookupPatchMeta strategicpatch.LookupPatchMeta
+		schema          oapi.Schema
+	)
+	// createPatchErrFormat := "creating patch with:\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s\nfor:"
+
+	// Create the versioned types from the type defined in the restmapping
+	// (which is the API version we'll be submitting the patch to)
+	versionedObject, err := scheme.Scheme.New(p.Mapping.GroupVersionKind)
+
+	switch {
```

Comment:
I know this part was reverted, however we could consider a refactor of the entire switch to be easier to read and maintain (we could have a separate story for this).